Sources/AppcuesKit/Presentation/Actions/Appcues/AppcuesRequestPushAction.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  AppcuesRequestPushAction.swift
+//  AppcuesKitTests
+//
+//  Created by Matt on 2024-02-26.
+//  Copyright © 2024 Appcues. All rights reserved.
+//
+
+import Foundation
+import UserNotifications
+
+@available(iOS 13.0, *)
+internal class AppcuesRequestPushAction: AppcuesExperienceAction {
+
+    static let type = "@appcues/request-push"
+
+    private weak var appcues: Appcues?
+
+    required init?(configuration: AppcuesExperiencePluginConfiguration) {
+        self.appcues = configuration.appcues
+    }
+
+    func execute(completion: @escaping ActionRegistry.Completion) {
+        guard let appcues = appcues else {
+            completion()
+            return
+        }
+
+        let resultHandler: (Bool, Error?) -> Void = { _, error in
+            if let error = error {
+                appcues.config.logger.error("@appcues/request-push failed with %{public}@", error.localizedDescription)
+            }
+
+            let pushMonitor = appcues.container.resolve(PushMonitoring.self)
+            let analyticsPublisher = appcues.container.resolve(AnalyticsPublishing.self)
+
+            pushMonitor.refreshPushStatus { _ in
+                analyticsPublisher.publish(TrackingUpdate(
+                    type: .event(name: Events.Device.deviceUpdated.rawValue, interactive: false),
+                    isInternal: true
+                ))
+
+                completion()
+            }
+        }
+
+        // Skip call to UNUserNotificationCenter.current() in tests to avoid crashing in package tests
+        #if DEBUG
+        guard ProcessInfo.processInfo.environment["XCTestBundlePath"] == nil else {
+            resultHandler(false, AppcuesTraitError(description: "Executing in test environment"))
+            return
+        }
+        #endif
+
+        let options: UNAuthorizationOptions = [.alert, .sound, .badge]
+        UNUserNotificationCenter.current().requestAuthorization(options: options, completionHandler: resultHandler)
```

Comment:
I forget, too much context switching, does this automatically trigger the token to get requested and sent back into the app delegate in didRegisterForRemoteNotificationsWithDeviceToken when permission is accepted?

Reply:
No, that's entirely separate and dependent on `UIApplication.registerForRemoteNotifications()`. I was wondering if we wanted to call that in this code, but I figured our install instructions would have that call in the startup sequence so it'd be redundant here.

Reply:
No, because `UIApplication.registerForRemoteNotifications()` can happen before the notification authorization. This is what `_pushBackgroundEnabled` would track. Then, once authorized here, `_pushEnabled = true`.

Reply:
ah right, thank you.